pkg/handler/common/provider/kubevirt.go

```diff
@@ -224,3 +231,87 @@ func KubeVirtStorageClassesWithClusterCredentialsEndpoint(ctx context.Context, u
 
 	return KubeVirtStorageClasses(ctx, kvKubeconfig)
 }
+
+func filterVMIPresets(vmiPresets apiv2.VirtualMachineInstancePresetList, quota kubermaticv1.MachineDeploymentVMResourceQuota) apiv2.VirtualMachineInstancePresetList {
+	filteredVMIPresets := apiv2.VirtualMachineInstancePresetList{}
+
+	// Range over the records and apply all the filters to each record.
+	// If the record passes all the filters, add it to the final slice.
+	for _, vmiPreset := range vmiPresets {
+		presetSpec := kubevirtv1.VirtualMachineInstancePresetSpec{}
+		if err := json.Unmarshal([]byte(vmiPreset.Spec), &presetSpec); err != nil {
+			continue
+		}
+
+		cpu, memory, err := GetKubeVirtPresetResourceDetails(presetSpec)
+		if err != nil {
+			continue
+		}
+
+		if handlercommon.FilterCPU(int(cpu.AsApproximateFloat64()), quota.MinCPU, quota.MaxCPU) && handlercommon.FilterMemory(int(memory.Value()/(1<<30)), quota.MinRAM, quota.MaxRAM) {
+			filteredVMIPresets = append(filteredVMIPresets, vmiPreset)
+		}
+	}
+	return filteredVMIPresets
+}
+
+// GetKubeVirtPresetResourceDetails extracts cpu and mem resource requests from the kubevirt preset
+// for CPU, take the value by priority:
+// - if resource limit is set, use that
+// - check if resource request is set, use that
+// - check if spec.cpu is set, if socket and threads are set then do the calculation, use that
+// for memory, take the value by priority:
+// - if resource limit is set, use that
+// - if resource request is set, use that.
+func GetKubeVirtPresetResourceDetails(presetSpec kubevirtv1.VirtualMachineInstancePresetSpec) (resource.Quantity, resource.Quantity, error) {
+	var err error
+	// Get CPU
+	cpuReq := resource.Quantity{}
+	if presetSpec.Domain.CPU != nil && (!presetSpec.Domain.Resources.Requests.Cpu().IsZero() || !presetSpec.Domain.Resources.Limits.Cpu().IsZero()) {
```

Comment:
the conditions we spoke about could have a more simple form since the function checks already some of them.
It could look like this

```
if presetSpec.Domain.CPU != nil {

		if !presetSpec.Domain.Resources.Requests.Cpu().IsZero() || !presetSpec.Domain.Resources.Limits.Cpu().IsZero() {
			// raise an error
		}

		cores := presetSpec.Domain.CPU.Cores
		if cores == 0 {
			cores = 1
		}
		threads := presetSpec.Domain.CPU.Threads
		if threads == 0 {
			threads = 1
		}
		sockets := presetSpec.Domain.CPU.Sockets
		if sockets == 0 {
			sockets = 1
		}

		cpuReq, err = resource.ParseQuantity(strconv.Itoa(int(cores * threads * sockets)))
		if err != nil {
			return resource.Quantity{}, resource.Quantity{}, fmt.Errorf("error parsing calculated KubeVirt VCPU: %w", err)
		}
	} else {
		if !presetSpec.Domain.Resources.Requests.Cpu().IsZero() {
			cpuReq = *presetSpec.Domain.Resources.Requests.Cpu()
		}

		if !presetSpec.Domain.Resources.Limits.Cpu().IsZero() {
			cpuReq = *presetSpec.Domain.Resources.Limits.Cpu()
		}
	}
```